yt/frontends/ahf/data_structures.py

```diff
@@ -0,0 +1,176 @@
+"""
+AHF data structures
+
+
+
+"""
+
+#-----------------------------------------------------------------------------
+# Copyright (c) 2017, yt Development Team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+#-----------------------------------------------------------------------------
+
+import glob
+import os
+import stat
+
+import numpy as np
+
+from yt.data_objects.static_output import \
+    Dataset, \
+    ParticleFile
+from yt.geometry.particle_geometry_handler import \
+    ParticleIndex
+from yt.utilities.cosmology import \
+    Cosmology
+
+from .fields import AHFHalosFieldInfo
+
+
+class AHFHalosFile(ParticleFile):
+    def __init__(self, ds, io, filename, file_id):
+        root, _ = os.path.splitext(filename)
+        candidates = glob.glob(root + '*.AHF_halos')
+        if len(candidates) == 1:
+            filename = candidates[0]
+        else:
+            raise ValueError('Too many AHF_halos files.')
+        names = self._read_column_names(filename)
+        self.data = np.genfromtxt(filename, names=names)
```

Comment:
I wouldn't worry too much about multiple reads for the same fields.  If you wanted to cache any fields, you could cache just the x, y, and z fields as they are used every single time to create the mask.  I think this could be converted easily to on-demand reading by having this `AHFHalosFile` object store the columns for each field name.  Then, in `_read_particle_fields`, you can use the `usecols` keyword with `np.genfromtxt` to get just the fields you want.